### PR TITLE
Add C++ wrappers to Prism's C API

### LINKED ISSUE #185: Speed up compile times by moving Prism translation out of `pipeline.cc`
All of our logic to translate the Prism AST into the Sorbet Parser AST lives in `pipeline.cc`, which is taking a while to compile.

https://github.com/Shopify/sorbet/blob/42119e9fc26e44f449c2116c8830814713f0f807/main/pipeline/pipeline.cc#L157-L561

It would be nice to move this into a standalone compilation unit, so it compiles faster. It would also just be better to have, from a code organization perspective.

### PR DESCRIPTION
**Pull request by amomchilov**
### Motivation

Closes #185. Best reviewed commit-by-commit.

This PR...

- Improves compilation times by not needing to recompile all of `pipeline.cc` whenever we're working on the translation layer.

    Here's the time taken to build the Prism test suite with:
    
    ```sh
    time ./bazel build //test:prism --config=dbg --define RUBY_PATH='/opt/rubies/3.3.4/' --test_output=errors
    ```
    
    |                           |    Time |
    |---------------------------|--------:|
    | No change (already built) |  0.314s |
    | Changing `Translator.cc`  |  4.522s |
    | Changing `pipeline.cc`    | 11.439s |

- Improves code organization, by grouping all prismatic stuff under `/parser/prism`
- Uses C++ memory management techniques to wrap the Prism C API's allocator and deallocator functions

That last point will be especially important when we start removing the translation layer, and these nodes start having much less definite lifetimes.

### Test plan

This is a pure refactor, and all the existing tests pass

### CODE REVIEW

File: parser/prism/Parser.cc [hunk 1]
```diff
@@ -0,0 +1,31 @@
+#include "parser/prism/parser.h"
+
+namespace sorbet::parser::Prism {
+
+pm_parser_t *Parser::tmp_public_get_raw_parser_pointer() {
+    return get_raw_parser_pointer();
+}
+
```
Comment (amomchilov) [inline parser/prism/Parser.cc:8]: Not necessary anymore?

```suggestion
```

File: parser/prism/Translator.cc [hunk 2]
```diff
@@ -0,0 +1,385 @@
+#include "Translator.h"
+
+template class std::unique_ptr<sorbet::parser::Node>;
+
+using namespace std;
+
```
Comment (amomchilov) [inline parser/prism/Translator.cc:6]: Bad.

```suggestion
```

**Review by egiurleo — APPROVED** [review]
This is great so far! I am super super on board with where this is headed.

I'm happy to talk about anything you need going forward --how best to merge this work, what else we need to move over to the new Parser wrapper class, etc. Let me know :)

**Review by KaanOzkan — APPROVED** [review]

**Review by egiurleo — APPROVED** [review]
This PR is * chef's kiss *. Thanks for making all the commits so readable. They really tell a story all together.

> Changing Translator.cc | 4.522s
> Changing pipeline.cc | 11.439s

200w

**Pull request merged**